ir/instr.cpp

```diff
@@ -1125,6 +1125,89 @@ unique_ptr<Instr> ExtractValue::dup(const string &suffix) const {
 }
 
 
+void InsertValue::addIdx(unsigned idx) {
+  idxs.emplace_back(idx);
+}
+
+vector<Value*> InsertValue::operands() const {
+  return { val, elt };
+}
+
+void InsertValue::rauw(const Value &what, Value &with) {
+  RAUW(val);
+  RAUW(elt);
+}
+
+void InsertValue::print(ostream &os) const {
+  os << getName() << " = insertvalue " << *val << ", " << *elt;
+  for (auto idx : idxs) {
+    os << ", " << idx;
+  }
+}
+
+StateValue update_repack(Type *type,
+                         const StateValue &val,
+                         const StateValue &elem,
+                         vector<unsigned> &indices) {
+  auto ty = type->getAsAggregateType();
+  unsigned cur_idx = indices.back();
+  indices.pop_back();
+  vector<StateValue> vals;
+  for (unsigned i = 0; i < ty->numElementsConst(); ++i) {
+    auto v = ty->extract(val, i);
+    if (i == cur_idx) {
+      vals.emplace_back(indices.empty() ?
+                        elem :
+                        update_repack(&ty->getChild(i), v, elem, indices));
+    } else {
+      vals.emplace_back(v);
+    }
+  }
+
+  return ty->aggregateVals(vals);
+}
+
+StateValue InsertValue::toSMT(State &s) const {
+  auto sv = s[*val];
+  auto elem = s[*elt];
+
+  Type *type = &val->getType();
+  vector<unsigned> idxs_reverse = idxs;
```

Comment:
vector has a constructor that can take iterators. so you can do vector(idxs.rbegin(), rend())